paramfetch.go

```diff
@@ -103,6 +107,37 @@ func (ft *fetch) maybeFetchAsync(ctx context.Context, name string, info paramFil
 		ft.fetchLk.Lock()
 		defer ft.fetchLk.Unlock()
 
+		var lockfail bool
+		var unlocker io.Closer
+		for {
+			unlocker, err = fslock.Lock(getParamDir(), lockFile)
+			if err == nil {
+				break
+			}
+
+			lockfail = true
+
+			le := fslock.LockedError("")
+			if xerrors.As(err, &le) {
+				log.Warnf("acquiring filesystem fetch lock: %s; will retry in %s", err, lockRetry)
+				time.Sleep(lockRetry)
+				continue
+			}
+			ft.errs = append(ft.errs, xerrors.Errorf("acquiring filesystem fetch lock: %w", err))
+			return
+		}
+		defer func() {
+			err := unlocker.Close()
+			if err != nil {
+				log.Errorw("unlock fs lock", "error", err)
+			}
+		}()
+		if lockfail {
+			// we've managed to get the lock, but we need to re-check file contents - maybe it's fetched now
+			ft.maybeFetchAsync(ctx, name, info)
```

Comment:
is this recursive call really better than just calling `checkFile` here and falling through to `doFetch` as needed?

Reply:
It's not really recursive as maybeFetchAsync runs this in a goroutine.

We're just calling maybeFetchAsync because:
* We don't want to be hashing files with those locks (want to do that in parallel, locks only for writing)
* Given the above, we'd want to drop those locks before calling checkFile
* After calling checkFile we'd need to acquire locks again
* So at that point we can just call maybeFetchAsync